native/explorer/src/local_message.rs

```diff
@@ -1,29 +1,48 @@
-use rustler::wrapper::{NIF_ENV, NIF_TERM};
-use rustler::{Env, Term};
-use std::ffi::c_void;
+use std::sync::Mutex;
 
-extern "C" {
-    pub fn local_message_open_resource(env: NIF_ENV) -> c_void;
-    fn local_message_on_gc(env: NIF_ENV, arg1: NIF_TERM, arg2: NIF_TERM) -> NIF_TERM;
-    fn is_local_message_when_gc(env: NIF_ENV, arg1: NIF_TERM) -> NIF_TERM;
+use rustler::env::SavedTerm;
+use rustler::{Env, LocalPid, OwnedEnv, Resource, ResourceArc, Term};
+
+pub struct LocalMessage {
+    pid: LocalPid,
+    env_and_term: Mutex<(OwnedEnv, SavedTerm)>,
 }
 
-#[rustler::nif]
-pub fn message_on_gc<'a>(env: Env<'a>, pid: Term<'a>, term: Term<'a>) -> Term<'a> {
-    unsafe {
-        Term::new(
-            env,
-            local_message_on_gc(env.as_c_arg(), pid.as_c_arg(), term.as_c_arg()),
-        )
+impl LocalMessage {
+    pub fn new(pid: LocalPid, term: Term<'_>) -> Self {
+        let env = OwnedEnv::new();
+        let term = env.save(term);
+        Self {
+            pid,
+            env_and_term: Mutex::new((env, term)),
```

Comment:
Is there an easy way to get the size of a data structure in Rust? My only concern is how much memory this uses, I am not very worried about the performance aspect since, as you said, it is never locked.

Reply:
There is a sizeof. But I think, in this particular case we should be able to get away with just `unsafe impl Sync` on the type.

Reply:
It depends on the size. I assume it is small compared to the env and term? But if it is order of magnitude larger, then I would go with unsafe, as we are sure it is safe in other ways.

Reply:
I think it's a `pthread` mutex, so ~40 bytes. Env and term are together probably around the same size. It's fine, I changed it to the `unsafe impl` now, much simpler anyway.